das/gcs_storage_service.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package das
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/das/dastree"
+	"github.com/offchainlabs/nitro/util/pretty"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+
+	googleStorage "cloud.google.com/go/storage"
+	flag "github.com/spf13/pflag"
+)
+
+type GCSStorageServiceConfig struct {
+	Enable                 bool   `koanf:"enable"`
+	Bucket                 string `koanf:"bucket"`
+	ObjectPrefix           string `koanf:"object-prefix"`
+	DiscardAfterTimeout    bool   `koanf:"discard-after-timeout"`
+	SyncFromStorageService bool   `koanf:"sync-from-storage-service"`
+	SyncToStorageService   bool   `koanf:"sync-to-storage-service"`
+}
+
+var DefaultGCSStorageServiceConfig = GCSStorageServiceConfig{}
+
+func GCSConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultGCSStorageServiceConfig.Enable, "enable storage/retrieval of sequencer batch data from an Google cloud storage bucket")
+	f.String(prefix+".bucket", DefaultGCSStorageServiceConfig.Bucket, "Google cloud storage bucket")
+	f.String(prefix+".object-prefix", DefaultGCSStorageServiceConfig.ObjectPrefix, "prefix to add to GCS objects")
+	f.Bool(prefix+".discard-after-timeout", DefaultGCSStorageServiceConfig.DiscardAfterTimeout, "discard data after its expiry timeout")
+	f.Bool(prefix+".sync-from-storage-service", DefaultGCSStorageServiceConfig.SyncFromStorageService, "enable gcs to be used as a source for regular sync storage")
+	f.Bool(prefix+".sync-to-storage-service", DefaultGCSStorageServiceConfig.SyncToStorageService, "enable gcs to be used as a sink for regular sync storage")
+}
+
+type GCSStorageService struct {
+	client              *googleStorage.Client
+	bucket              string
+	objectPrefix        string
+	discardAfterTimeout bool
+}
+
+func NewGCSStorageService(config GCSStorageServiceConfig) (StorageService, error) {
+	// https://cloud.google.com/docs/authentication/provide-credentials-adc
+	client, err := googleStorage.NewClient(context.TODO())
```

Comment:
You can make `NewGCSStorageService` take a `Context` so that you can pass it to `NewClient` rather than using TODO